sqlglotrs/src/tokenizer.rs

```diff
@@ -0,0 +1,642 @@
+        keyword_trie.add(settings.quotes.keys().filter(trie_filter));
+        keyword_trie.add(settings.format_strings.keys().filter(trie_filter));
+
+        Tokenizer {
+            settings,
+            keyword_trie,
+        }
+    }
+
+    pub fn tokenize(&self, sql: &str) -> Vec<Token> {
+        let mut state = TokenizerState::new(sql, &self.settings, &self.keyword_trie);
+        state.tokenize()
+    }
+}
+
+#[derive(Debug)]
+struct TokenizerState<'a> {
+    sql: Vec<char>,
+    size: usize,
+    tokens: Vec<Token>,
+    start: usize,
+    current: usize,
+    line: usize,
+    column: usize,
+    comments: Vec<String>,
+    is_end: bool,
+    current_char: char,
+    peek_char: char,
+    previous_token_line: Option<usize>,
+    keyword_trie: &'a Trie,
+    settings: &'a TokenizerSettings,
+}
+
+impl<'a> TokenizerState<'a> {
+    fn new(
+        sql: &str,
+        settings: &'a TokenizerSettings,
+        keyword_trie: &'a Trie,
+    ) -> TokenizerState<'a> {
+        TokenizerState {
+            sql: sql.chars().collect(),
+            size: sql.len(),
+            tokens: Vec::new(),
+            start: 0,
+            current: 0,
+            line: 1,
+            column: 0,
+            comments: Vec::new(),
+            is_end: false,
+            current_char: '\0',
+            peek_char: '\0',
+            previous_token_line: None,
+            keyword_trie,
+            settings,
+        }
+    }
+
+    fn tokenize(&mut self) -> Vec<Token> {
+        self.scan(None);
+        std::mem::replace(&mut self.tokens, Vec::new())
+    }
+
+    fn scan(&mut self, until_peek_char: Option<char>) {
+        while self.size > 0 && !self.is_end {
+            self.start = self.current;
+            self.advance(1, false, false);
+
+            if self.current_char == '\0' {
+                break;
+            }
+
+            if !self.settings.white_space.contains_key(&self.current_char) {
+                if self.current_char.is_digit(10) {
+                    self.scan_number();
+                } else {
+                    match self.settings.identifiers.get(&self.current_char) {
+                        Some(identifier_end) => self.scan_identifier(&identifier_end.to_string()),
+                        None => self.scan_keyword(),
+                    }
+                }
+            }
+
+            match until_peek_char {
+                Some(c) => {
+                    if self.peek_char == c {
+                        break;
+                    }
+                }
+                None => (),
+            }
+        }
+        if !self.tokens.is_empty() && !self.comments.is_empty() {
+            self.tokens
+                .last_mut()
+                .unwrap()
+                .append_comments(&mut self.comments);
+        }
+    }
+
+    fn advance(&mut self, i: usize, alnum: bool, backtrack: bool) {
+        match self.settings.white_space.get(&self.current_char) {
```

Comment:
I feel like the `if let ... else` syntax is better for these little checks.
Seems more idiomatic too: https://rust-lang.github.io/rfcs/0160-if-let.html

Reply:
TIL, thanks